<a href="https://colab.research.google.com/github/animesh-11/AI_ML/blob/main/Linear_Regression_Assessment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Load the "College.xlsx" dataset, remove the column containing university names, calculate and round descriptive statistics for numeric columns to 2 decimal places, one-hot encode the 'Private' variable with `drop_first=True`, define 'Grad.Rate' as the target variable, split the data into 70% training and 30% testing sets using `random_state=42`, fit a regression model, output the model coefficients rounded to 2 decimal places, obtain predictions for both training and testing sets, and calculate and round R², Mean Squared Error, and Mean Absolute Error for both training and testing sets to 2 decimal places.

## data_loading

### Subtask:
Load the `College.xlsx` dataset using pandas and remove the column containing university names.


**Reasoning**:
I need to import the pandas library, load the 'College.xlsx' dataset, and display its head to identify the column containing university names as per the instructions.



In [1]:
import pandas as pd

df = pd.read_excel('/content/College.xlsx')
print("Original DataFrame head:")
print(df.head())

Original DataFrame head:
                     Unnamed: 0 Private  Apps  Accept  Enroll  Top10perc  \
0  Abilene Christian University     Yes  1660    1232     721         23   
1            Adelphi University     Yes  2186    1924     512         16   
2                Adrian College     Yes  1428    1097     336         22   
3           Agnes Scott College     Yes   417     349     137         60   
4     Alaska Pacific University     Yes   193     146      55         16   

   Top25perc  F.Undergrad  P.Undergrad  Outstate  Room.Board  Books  Personal  \
0         52         2885          537      7440        3300    450      2200   
1         29         2683         1227     12280        6450    750      1500   
2         50         1036           99     11250        3750    400      1165   
3         89          510           63     12960        5450    450       875   
4         44          249          869      7560        4120    800      1500   

   PhD  Terminal  S.F.Ratio  pe

**Reasoning**:
The previous step identified 'Unnamed: 0' as the column containing university names. Now, I will remove this column from the DataFrame and display the head to confirm the change.



In [2]:
df = df.drop('Unnamed: 0', axis=1)
print("DataFrame head after removing 'Unnamed: 0' column:")
print(df.head())

DataFrame head after removing 'Unnamed: 0' column:
  Private  Apps  Accept  Enroll  Top10perc  Top25perc  F.Undergrad  \
0     Yes  1660    1232     721         23         52         2885   
1     Yes  2186    1924     512         16         29         2683   
2     Yes  1428    1097     336         22         50         1036   
3     Yes   417     349     137         60         89          510   
4     Yes   193     146      55         16         44          249   

   P.Undergrad  Outstate  Room.Board  Books  Personal  PhD  Terminal  \
0          537      7440        3300    450      2200   70        78   
1         1227     12280        6450    750      1500   29        30   
2           99     11250        3750    400      1165   53        66   
3           63     12960        5450    450       875   92        97   
4          869      7560        4120    800      1500   76        72   

   S.F.Ratio  perc.alumni  Expend  Grad.Rate  
0       18.1           12    7041         60  
1

## descriptive_statistics

### Subtask:
Calculate and round descriptive statistics for numeric columns to 2 decimal places.


**Reasoning**:
I need to calculate descriptive statistics for all numeric columns in the DataFrame `df` and then round the results to two decimal places, as specified in the subtask.



In [3]:
descriptive_stats = df.describe().round(2)
print("Descriptive Statistics (rounded to 2 decimal places):")
print(descriptive_stats)

Descriptive Statistics (rounded to 2 decimal places):
           Apps    Accept   Enroll  Top10perc  Top25perc  F.Undergrad  \
count    777.00    777.00   777.00     777.00      777.0       777.00   
mean    3001.64   2018.80   779.97      27.56       55.8      3699.91   
std     3870.20   2451.11   929.18      17.64       19.8      4850.42   
min       81.00     72.00    35.00       1.00        9.0       139.00   
25%      776.00    604.00   242.00      15.00       41.0       992.00   
50%     1558.00   1110.00   434.00      23.00       54.0      1707.00   
75%     3624.00   2424.00   902.00      35.00       69.0      4005.00   
max    48094.00  26330.00  6392.00      96.00      100.0     31643.00   

       P.Undergrad  Outstate  Room.Board    Books  Personal     PhD  Terminal  \
count       777.00    777.00      777.00   777.00    777.00  777.00    777.00   
mean        855.30  10440.67     4357.53   549.38   1340.64   72.66     79.70   
std        1522.43   4023.02     1096.70   16

## feature_engineering

### Subtask:
One-hot encode the 'Private' variable using `pandas.get_dummies()` with `drop_first=True`.


**Reasoning**:
I need to one-hot encode the 'Private' column using `pd.get_dummies()` with `drop_first=True`, then concatenate the results back to the original DataFrame and drop the original 'Private' column, and finally display the head of the updated DataFrame.



In [4]:
private_dummies = pd.get_dummies(df['Private'], drop_first=True, dtype=int)
df = pd.concat([df, private_dummies], axis=1)
df = df.drop('Private', axis=1)
print("DataFrame head after one-hot encoding 'Private' and dropping original column:")
print(df.head())

DataFrame head after one-hot encoding 'Private' and dropping original column:
   Apps  Accept  Enroll  Top10perc  Top25perc  F.Undergrad  P.Undergrad  \
0  1660    1232     721         23         52         2885          537   
1  2186    1924     512         16         29         2683         1227   
2  1428    1097     336         22         50         1036           99   
3   417     349     137         60         89          510           63   
4   193     146      55         16         44          249          869   

   Outstate  Room.Board  Books  Personal  PhD  Terminal  S.F.Ratio  \
0      7440        3300    450      2200   70        78       18.1   
1     12280        6450    750      1500   29        30       12.2   
2     11250        3750    400      1165   53        66       12.9   
3     12960        5450    450       875   92        97        7.7   
4      7560        4120    800      1500   76        72       11.9   

   perc.alumni  Expend  Grad.Rate  Yes  
0        

## data_splitting

### Subtask:
Define 'Grad.Rate' as the target variable and split the data into 70% training and 30% testing sets using `sklearn.model_selection.train_test_split` with `random_state=42`.


**Reasoning**:
I need to import the `train_test_split` function, define the target variable and features, and then split the data into training and testing sets as specified in the instructions. Printing the shapes of the resulting datasets will confirm the split was successful.



In [5]:
from sklearn.model_selection import train_test_split

y = df['Grad.Rate']
X = df.drop('Grad.Rate', axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

print("Shapes of the datasets after splitting:")
print(f"X_train shape: {X_train.shape}")
print(f"X_test shape: {X_test.shape}")
print(f"y_train shape: {y_train.shape}")
print(f"y_test shape: {y_test.shape}")

Shapes of the datasets after splitting:
X_train shape: (543, 17)
X_test shape: (234, 17)
y_train shape: (543,)
y_test shape: (234,)


## model_training

### Subtask:
Fit a regression model (e.g., Linear Regression) on the training data.


**Reasoning**:
I need to import the `LinearRegression` model from `sklearn.linear_model`, instantiate it, and then fit it to the training data `X_train` and `y_train`.



In [6]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

print("Linear Regression model fitted successfully.")

Linear Regression model fitted successfully.


## coefficient_output

### Subtask:
Output the model coefficients, rounded to 2 decimal places.


**Reasoning**:
I need to extract the coefficients from the fitted model, associate them with their respective feature names, round them to two decimal places, and then print them.



In [7]:
coefficients = pd.Series(model.coef_, index=X_train.columns).round(2)
print("Model Coefficients (rounded to 2 decimal places):")
print(coefficients)

Model Coefficients (rounded to 2 decimal places):
Apps           0.00
Accept        -0.00
Enroll         0.00
Top10perc      0.03
Top25perc      0.14
F.Undergrad   -0.00
P.Undergrad   -0.00
Outstate       0.00
Room.Board     0.00
Books         -0.00
Personal      -0.00
PhD            0.14
Terminal      -0.11
S.F.Ratio     -0.03
perc.alumni    0.34
Expend        -0.00
Yes            5.05
dtype: float64


## model_prediction

### Subtask:
Obtain predictions for both training and testing sets using the fitted model.

## model_prediction

### Subtask:
Obtain predictions for both training and testing sets using the fitted model.

**Reasoning**:
Now that the model coefficients have been outputted, the next logical step is to use the fitted model to generate predictions on both the training and testing datasets.



In [8]:
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

print("Predictions generated for training and testing sets.")

Predictions generated for training and testing sets.


## model_evaluation

### Subtask:
Calculate and round R², Mean Squared Error, and Mean Absolute Error for both training and testing sets to 2 decimal places.

## model_evaluation

### Subtask:
Calculate and round R², Mean Squared Error, and Mean Absolute Error for both training and testing sets to 2 decimal places.

**Reasoning**:
Now that the predictions are available, I need to calculate the R², Mean Squared Error (MSE), and Mean Absolute Error (MAE) for both the training and testing sets, rounding each metric to two decimal places, as specified in the subtask.



In [11]:
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

# Calculate metrics for training set
r2_train = round(r2_score(y_train, y_train_pred), 2)
mse_train = round(mean_squared_error(y_train, y_train_pred), 2)
mae_train = round(mean_absolute_error(y_train, y_train_pred), 2)

# Calculate metrics for testing set
r2_test = round(r2_score(y_test, y_test_pred), 2)
mse_test = round(mean_squared_error(y_test, y_test_pred), 2)
mae_test = round(mean_absolute_error(y_test, y_test_pred), 2)

print("Model Evaluation Metrics:")
print(f"  Training R²: {r2_train}")
print(f"  Training MSE: {mse_train}")
print(f"  Training MAE: {mae_train}")
print(f"  Testing R²: {r2_test}")
print(f"  Testing MSE: {mse_test}")
print(f"  Testing MAE: {mae_test}")

# Get and print the intercept term
intercept = round(model.intercept_, 2)
print(f"\nIntercept Term (rounded to 2 decimal places): {intercept}")

Model Evaluation Metrics:
  Training R²: 0.46
  Training MSE: 160.59
  Training MAE: 9.37
  Testing R²: 0.44
  Testing MSE: 160.48
  Testing MAE: 9.6

Intercept Term (rounded to 2 decimal places): 34.94


**Reasoning**:
The previous code failed because the `.round(2)` method was incorrectly applied directly to the output of `r2_score`, `mean_squared_error`, and `mean_absolute_error`, which are float values. I need to use the built-in `round()` function on the results of these metrics.



In [10]:
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

# Calculate metrics for training set
r2_train = round(r2_score(y_train, y_train_pred), 2)
mse_train = round(mean_squared_error(y_train, y_train_pred), 2)
mae_train = round(mean_absolute_error(y_train, y_train_pred), 2)

# Calculate metrics for testing set
r2_test = round(r2_score(y_test, y_test_pred), 2)
mse_test = round(mean_squared_error(y_test, y_test_pred), 2)
mae_test = round(mean_absolute_error(y_test, y_test_pred), 2)

print("Model Evaluation Metrics:")
print(f"  Training R²: {r2_train}")
print(f"  Training MSE: {mse_train}")
print(f"  Training MAE: {mae_train}")
print(f"  Testing R²: {r2_test}")
print(f"  Testing MSE: {mse_test}")
print(f"  Testing MAE: {mae_test}")

Model Evaluation Metrics:
  Training R²: 0.46
  Training MSE: 160.59
  Training MAE: 9.37
  Testing R²: 0.44
  Testing MSE: 160.48
  Testing MAE: 9.6


## Summary:

### Data Analysis Key Findings

*   The `College.xlsx` dataset was successfully loaded, and the non-numeric 'Unnamed: 0' column, which contained university names, was removed.
*   Descriptive statistics for all numeric columns were calculated and rounded to 2 decimal places.
*   The 'Private' categorical variable was successfully one-hot encoded with `drop_first=True`, resulting in a new 'Yes' column representing private universities (1 for private, 0 for public).
*   The dataset was split into training (543 samples, 70%) and testing (234 samples, 30%) sets, with 'Grad.Rate' defined as the target variable and `random_state=42` used for reproducibility.
*   A Linear Regression model was trained on the processed training data.
*   The model coefficients were calculated and rounded to 2 decimal places. For example, 'perc.alumni' had a coefficient of 0.34, and the 'Yes' variable (indicating a private university) had a coefficient of 5.05.
*   Model predictions were generated for both the training and testing sets.
*   The model's performance was evaluated using R², Mean Squared Error (MSE), and Mean Absolute Error (MAE), all rounded to 2 decimal places:
    *   **Training Set**: R² = 0.46, MSE = 160.59, MAE = 9.37.
    *   **Testing Set**: R² = 0.44, MSE = 160.48, MAE = 9.60.

### Insights or Next Steps

*   The model explains approximately 44-46% of the variance in graduation rates, indicating that while it captures some patterns, a significant portion remains unexplained. The testing R² (0.44) is very close to the training R² (0.46), suggesting that the model generalizes well to unseen data without significant overfitting.
*   Considering the MAE of around 9.60 on the test set, the model's predictions for graduation rates are, on average, off by approximately 9.6 percentage points. Further analysis could explore more complex models, feature engineering (e.g., interaction terms, polynomial features), or feature selection to potentially improve predictive performance.
